src/Avalonia.Animation/Keyframes/KeyFramesStateMachine`1.cs

```diff
@@ -0,0 +1,276 @@
+
+                case KeyFramesStates.RUN_FORWARDS:
+
+                    if (_durationFrameCount > _durationTotalFrameCount)
+                    {
+                        _currentState = KeyFramesStates.RUN_COMPLETE;
+                        goto checkstate;
+                    }
+
+                    _tempDuration = (double)_durationFrameCount / _durationTotalFrameCount;
+                    _currentState = KeyFramesStates.RUN_APPLYVALUE;
+
+                    goto checkstate;
+
+                case KeyFramesStates.RUN_BACKWARDS:
+
+                    if (_durationFrameCount > _durationTotalFrameCount)
+                    {
+                        _currentState = KeyFramesStates.RUN_COMPLETE;
+                        goto checkstate;
+                    }
+
+                    _tempDuration = (double)(_durationTotalFrameCount - _durationFrameCount) / _durationTotalFrameCount;
+                    _currentState = KeyFramesStates.RUN_APPLYVALUE;
+
+                    goto checkstate;
+
+                case KeyFramesStates.RUN_APPLYVALUE:
+
+                    _easedTime = _targetAnimation.Easing.Ease(_tempDuration);
+
+                    _durationFrameCount++;
+                    _lastInterpValue = Interpolator(_easedTime);
+                    _targetObserver.OnNext(_lastInterpValue);
+                    _currentState = KeyFramesStates.DO_RUN;
+
+                    break;
+
+                case KeyFramesStates.RUN_COMPLETE:
+
+                    if (_checkLoopAndRepeat)
+                    {
+                        _delayFrameCount = 0;
+                        _durationFrameCount = 0;
+
+                        if (_isLooping)
+                        {
+                            _currentState = KeyFramesStates.DO_RUN;
+                        }
+                        else if (_isRepeating)
+                        {
+                            if (_currentIteration >= _repeatCount)
+                            {
+                                _currentState = KeyFramesStates.STOP;
+                            }
+                            else
+                            {
+                                _currentState = KeyFramesStates.DO_RUN;
+                            }
+                            _currentIteration++;
+                        }
+
+                        if (_animationDirection == PlaybackDirection.Alternate
+                         || _animationDirection == PlaybackDirection.AlternateReverse)
+                            _isReversed = !_isReversed;
+
+                        break;
+                    }
+
+                    _currentState = KeyFramesStates.STOP;
+                    goto checkstate;
+
+                case KeyFramesStates.STOP:
+                    if (_fillMode == FillMode.Forward
+                     || _fillMode == FillMode.Both)
+                    {
+                        _targetControl.SetValue(_parent.Property, _lastInterpValue, BindingPriority.Animation);
```

Comment:
Hmm, should this be set with `Animtion` or `LocalValue` priority? I'm not sure...